rbd-mirror: daemon skeleton

jdurgin opened the pull request:
This is a non-functional draft of the structure for a daemon to mirror
changes from rbd images in one cluster to other clusters. There are
many missing pieces (like the part that actually replays the changes)
and optimizations (some of which may be necessary features) that
we should think about. I'll elaborate on these tomorrow.

This is just an initial stab at a basic structure, posted for comment.

jdurgin commented:
Some future work for mirroring on top of this:

##### Optimizations

* index of mirrored images that can be watched for changes instead of
  polling all images' flags
* convert to aio state machines
* use librbd internals rather than librbd itself (this is probably
  necessary for image replay anyway)

##### Features

* peer cluster liveness detection
* mirroring configuration versioning, so we can tell which cluster is
  an image should be active in
* For mirroring enabled on existing images with incomplete journals,
  take an internal snapshot as a starting point and clean it up when
  deleting the image
* HA rbd-mirror daemons - distribute work among many daemons - perhaps
  deterministically decide images to mirror based on client id in the
  local cluster, and number of rbd-mirror instances. Use watch/notify
  to check daemon liveness.
* ability to configure per-image peers (overriding pool settings?) not
  sure this is needed. enabling/disabling mirror per-image may be good
  enough

yuyuyu101 commented on src/tools/rbd_mirror/ImageWatcher.cc:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  librbd::RBD rbd;
+  for (auto kv : pools) {
```
do you think it will be too heavy? Do we need something like RBD_DIRECTORY to store mirroring or alike?

yuyuyu101 replied:
oh, ignore this.... I see the later comment

yuyuyu101 commented:
If a cluster's image want to setup a another cluster's image as the mirror image without , do we have something to limit this like quota? Maybe I'm missing what already implemented.

dillaman commented on src/librbd/internal.cc:
```diff
@@ -1998,6 +2024,19 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
     return ictx->get_flags(ictx->snap_id, flags);
   }
 
+  int set_flags(ImageCtx *ictx, uint64_t flags, bool enabled)
```
Rename to ```update_flags``` since set_flags implies you are enabling all flags passed as the argument and disabling those not.  Also probably should prevent the API from updating ```RBD_FLAG_OBJECT_MAP_INVALID``` and ```RBD_FLAG_FAST_DIFF_INVALID```.

dillaman commented on src/tools/rbd_mirror/PoolWatcher.cc:
```diff
@@ -0,0 +1,143 @@
+
+void PoolWatcher::refresh_pools()
+{
+  dout(20) << __func__ << dendl;
```
Why ```dout``` and ```derr``` instead of ```ldout``` and ```lderr```?

jdurgin replied:
as a standalone binary, we can use the global version that relies on g_ceph_context instead of passing cct everywhere like we need to with library version of dout (ldout/lderr)

dillaman commented on src/tools/rbd_mirror/Mirror.cc:
```diff
@@ -0,0 +1,114 @@
+  }
+}
+
+void Mirror::update_replayers(map<peer_t, set<int64_t> > peer_configs,
```
Pass by const ref -- could have thousands of images

jdurgin replied:
yes, need to be extra careful will all these potentially large structures

dillaman commented on src/tools/rbd_mirror/ImageWatcher.cc:
```diff
@@ -0,0 +1,151 @@
+  m_timer.shutdown();
+}
+
+map<int64_t, set<string> > ImageWatcher::get_images() const
```
Already own the lock -- should this just return a const ref?

dillaman commented on src/tools/rbd_mirror/Mirror.cc:
```diff
@@ -0,0 +1,114 @@
+  for (auto it = m_replayers.begin(); it != m_replayers.end();) {
+    peer_t peer = it->first;
+    if (peers.find(peer) == peers.end()) {
+      m_replayers.erase(it++);
```
Do you need to shut down the replayer?

jdurgin replied:
handled in the destructor now

dillaman commented on src/tools/rbd_mirror/Replayer.cc:
```diff
@@ -0,0 +1,125 @@
+  m_remote->shutdown();
+}
+
+void Replayer::set_sources(const map<int64_t, set<string> > &images)
```
I think this is backwards -- unless the idea is for the mirroring daemon to push updates to the remote.  You scan your local pools to know which pools are potentially mirrrored.   Then, for each remote pool, you would scan those pools for images that you need to sync/pull to the local pool.

This gets me thinking about how we should handle image removals -- and how to propagate those changes.  If we had an image being mirrored and we disable mirroring within one cluster, should the image be erased from the peers?  If the user deletes a mirrorred image from one cluster, should the image be erased from all peers?  Sounds like we almost need a journal for ```rbd_directory``` to track changes like that.  

And given that, should we just move the mirror enabled flag to the ```rbd_directory``` itself instead of storing it in each image?

jdurgin replied:
I was writing this assuming a push model, but can change to pull instead. I don't think there was a good reason for thinking of it as push - there's not much local sharing that would be useful. Now that I think about failover scenarios, pull does seem simpler.

For image removals, I was thinking we could treat them like [dis|en]abling mirroring on an image (and send a notify when we do that), but more and more it seems like storing the mirroring state in ``rbd_directory`` would be better than using generic image flags. I was thinking deletion would be mirrored as well.

This reminds me of another thing I don't think we've discussed:
Should ``rbd-mirror`` keep mirrored pools strictly in sync among clusters, or only operate on images with mirroring enabled? i.e., if non-mirrored images are created in a secondary cluster, should we do anything (warn, delete), or ignore them? I've been thinking the latter in both cases, but if a new mirrored image is created in the primary cluster, and the secondary cluster has one with the same name (without mirroring enabled), we have a problem. Perhaps in this kind of case we should expose a warning as part of the mirroring status, so the admin can decide what to do.

Another aspect is preventing accidental modification of mirrored images in a secondary cluster. It seems we'd want to disable writes to them at least, and perhaps only allow changing mirroring state (enabling/disabling) in the primary cluster.

we may want a way to prevent mirroring from being enabled for images in a secondary cluster.

dillaman replied:
It certainly is less work / less state tracking if we just say that is mirroring is enabled on a pool and if journaling is enabled on an image, it should be mirrored.  But if we need to track deletion status, I don't know if it saves that much effort.

WRT to designating primary / secondary clusters, it certainly can add some additional sanity safety checks.  i.e. prevent image creation, deletion, modification, etc on a secondary cluster pool.  Failover, especially in an automated, split-brain situation becomes a little more troublesome I think, though.  In an OpenStack environment, is there a standard way to initiate a failover to a new data center or are ppl required to roll their own solutions?  

An alternative approach is to track epochs within each image (or associated journal).  When an image acquires the exclusive lock, it starts a new epoch.  For the RBD mirroring daemon, we could have it be the remote journal reader and have it directly re-write the event entries in the local journal (while keeping the associated image open via librbd in a special live replay mode of operation where it reuses the existing replayer logic to apply the journal events).  Another alternative is a new API method to pass an opaque journal event to librbd image for it to directly apply (basically don't want to re-implement replay logic in rbd-mirror).  The mirroring daemon would know it should be pulling data from a peer if the peer has the highest epoch number (technically it should go sequentially between all known peers to copy epoch events in-order).   

We can also keep track of a period to handle the case of an unclean transition between clusters.  On rbd_close, we can record a "safely closed" event in the journal.  If that event is missing when acquiring the exclusive lock and replaying a peer's journal (i.e. after a cluster transition), we know that we need to start a new period.  This in effect would allow the peer to detect that it is now inconsistent with the local copy (and would require intervention by someone to pick the real copy).  This might require some interlocks between the local rbd-mirror and the librbd client to ensure all available remote events have been safely recorded locally.

I like the idea of a special file that can hold all the known per-image errors / state for rbd-mirror.  Much cleaner than attempting to scan the logs for error state and easy to integrate with the storage manager.

trociny commented:
I think the case when some images from cluster1/pool are mirrored to cluster2/pool while in the same time some other images are mirrored from cluster2/pool to cluster1/pool (i.e. for some images cluster1/pool is primary and cluster2/pool is secondary, while for others it is vice verse) could be useful. E.g. it would allow to balance load between datacenters, (live) migrating VMs from one DC to another.

Are we going to support such case or whould it be a real mess?

trociny commented:
Just a note, that failovering will be a more complicated if we have several secondaries. A secondary that remains secondary, before switching to the new master, should synchronize its commit position
with the new master. If the old master is still alive, it would be possible by comparing the clients commit positions in the master journal, but this won't be always the case, so the secondaries should
keep enough info to synchronize themselves. E.g. journaling is also enabled on the secondary images, the mirroring daemon pulls and applies journal events from the master, updating the client commit
position in the master, but also registers and updates the position on all other secondaries?

dillaman commented:
@trociny Agreed -- the mirroring daemon will need to apply events sequentially by epoch. If the local image becomes primary, that should imply that the mirroring daemon has transferred the "image safely closed" event from a peer or the interlock timed out and a new period should be started (e.g. possible split brain).

jdurgin commented:
changed to a pull model, added some unit tests, a basic man page and packaging. Now using image ids and librbd internals instead of names and the public interface.
